Add Test Cases and Update Package.md (#2)
ballerina-platform/module-ballerinax-hubspot.crm.properties: README.md

         // Parse and print the response
         json responseJson = check tokenResponse.getJsonPayload();
         io:println("Access and Refresh Token Response: ", responseJson);
      }
   ```

   - Linux/macOS

     ```bash
     curl --request POST \
     --url https://api.hubapi.com/oauth/v1/token \
     --header 'content-type: application/x-www-form-urlencoded' \
     --data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
     ```

   - Windows

     ```bash
     curl --request POST ^
     --url https://api.hubapi.com/oauth/v1/token ^
     --header 'content-type: application/x-www-form-urlencoded' ^
     --data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
     ```

   This command will return the access token necessary for API calls.

   ```json
   {
     "token_type": "bearer",
     "refresh_token": "<Refresh Token>",
     "access_token": "<Access Token>",
     "expires_in": 1800
   }
   ```

5. Store the access token securely for use in your application.

## Quickstart

> To use the `HubSpot Properties connector` in your Ballerina application, update the `.bal` file as follows:

Reviewer: ```suggestion
To use the `HubSpot CRM Properties connector` in your Ballerina application, update the `.bal` file as follows:
```

### Step 1: Import the module

> Import the `hubspot.properties` module and `oauth2` module.
> 
> ```ballerina
> import ballerinax/hubspot.properties as hsproperties;
> import ballerina/oauth2;

Reviewer: ```suggestion
Import the `hubspot.crm.properties` module and `oauth2` module.

```ballerina
import ballerinax/hubspot.crm.properties as hsproperties;
import ballerina/oauth2;
```

```
### Step 2: Instantiate a new connector

1. Instantiate a `hsproperties:ConnectionConfig` with the obtained credentials and initialize the connector with it.

    ```ballerina 
    configurable string clientId = ?;
    configurable string clientSecret = ?;
    configurable string refreshToken = ?;

>     final hsproperties:ConnectionConfig config = {
>         auth : {
>             clientId,
>             clientSecret,
>             refreshToken,
>             credentialBearer: oauth2:POST_BODY_BEARER
>         }
>     };
> 
>     final hsproperties:Client hsproperties = check new (config);

Reviewer: ```suggestion
    hsproperties:ConnectionConfig config = {
        auth : {
            clientId,
            clientSecret,
            refreshToken,
            credentialBearer: oauth2:POST_BODY_BEARER
        }
    };
    final hsproperties:Client hubSpot = check new (config);
```

    ```

2. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Create a Property Group

```ballerina
public function main() returns error? {

>    hsproperties:PropertyGroupCreate propertyGroupInput = { "name": "examplePropertyGroup",

Reviewer: Shall we format this code properly ?

Author: done

                                           "displayOrder": -1,
                                           "label": "This is an example Property Group"};
                                           

>    hsproperties:PropertyGroup response = check hubspot->/[testObjectType]/groups.post(payload = propertyGroupInput);

Reviewer: ```suggestion
   hsproperties:PropertyGroup response = check hubSpotProperties->/[testObjectType]/groups.post(payload = propertyGroupInput);
```

}
```

#### Run the Ballerina application

```bash
bal run
```

## Examples

The `Ballerina HubSpot CRM Properties Connector` connector provides practical examples illustrating usage in various scenarios. Explore these [examples](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/tree/8dfd490047721c5e72f0a8d995636860769e3c82/examples), covering the following use cases:

> 1. [Customer Behavior Handling](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/tree/8dfd490047721c5e72f0a8d995636860769e3c82/examples/customer-behavior)
> 2. [Marketing Preference Management](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/tree/8dfd490047721c5e72f0a8d995636860769e3c82/examples/marketing-preference)

Reviewer: Check for other places as well.
```suggestion
1. [Customer Behavior Handling](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/tree/main/examples/customer-behavior)
2. [Marketing Preference Management](https://github.com/ballerina-platform/module-ballerinax-hubspot.crm.properties/tree/main/examples/marketing-preference)
```

## Overview

[HubSpot ](https://www.hubspot.com/) is an AI-powered customer relationship management (CRM) platform. 

> The `ballerinax/hubspot.crm.properties` offers APIs to connect and interact with the [Properties](https://developers.hubspot.com/docs/reference/api/crm/properties)  endpoints , specifically based on the [API v3 documention](https://developers.hubspot.com/docs/reference/api/crm/properties)

Reviewer 2: ```suggestion
The `ballerinax/hubspot.crm.properties` offers APIs to connect and interact with the [Properties](https://developers.hubspot.com/docs/reference/api/crm/properties) endpoints, specifically based on the [API v3 documention](https://developers.hubspot.com/docs/reference/api/crm/properties)
```

### Step 2 (Optional): Create a [Developer Test Account](https://developers.hubspot.com/beta-docs/getting-started/account-types#developer-test-accounts) under your account

Within app developer accounts, you can create developer test accounts to test apps and integrations without affecting any real HubSpot data.

**Note: These accounts are only for development and testing purposes. In production you should not use Developer Test Accounts.**

1. Go to Test Account section from the left sidebar.  
   ![Hubspot developer portal](https://github.com/ManadaHerath/module-ballerinax-hubspot.crm.properties/blob/6a36221acc4771c608bcd8951bedca75b3e21024/docs/resources/test_acc_1.png)

2. Click Create developer test account.  
   ![Hubspot developer testacc](https://github.com/ManadaHerath/module-ballerinax-hubspot.crm.properties/blob/6a36221acc4771c608bcd8951bedca75b3e21024/docs/resources/test_acc_2.png)

3. In the dialogue box, give a name to your test account and click create.  
   ![Hubspot developer testacc3](https://github.com/ManadaHerath/module-ballerinax-hubspot.crm.properties/blob/6a36221acc4771c608bcd8951bedca75b3e21024/docs/resources/test_acc_3.png)

> 

Reviewer 2: ```suggestion
```

- Navigate to the Auth section of your app. Make sure to save the provided Client ID and Client Secret.  
  ![Hubspot get credentials](https://github.com/ManadaHerath/module-ballerinax-hubspot.crm.properties/blob/6f8800ecab2b9e4173d466b42d17805457fbc6d2/docs/resources/get_credentials.png)

### Step 6: Setup Authentication Flow

Before proceeding with the Quickstart, ensure you have obtained the Access Token using the following steps:

1. Create an authorization URL using the following format:

   ```
   https://app.hubspot.com/oauth/authorize?client_id=<YOUR_CLIENT_ID>&scope=<YOUR_SCOPES>&redirect_uri=<YOUR_REDIRECT_URI>
   ```

   Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI>` and `<YOUR_SCOPES>` with your specific value.

2. Paste it in the browser and select your developer test account to intall the app when prompted.
3. A code will be displayed in the browser. Copy the code.

   ```
   Received code: na1-129d-860c-xxxx-xxxx-xxxxxxxxxxxx
   ```

4. Run the following curl command. Replace the `<YOUR_CLIENT_ID>`, `<YOUR_REDIRECT_URI`> and `<YOUR_CLIENT_SECRET>` with your specific value. Use the code you received in the above step 3 as the `<CODE>`.

   - Linux/macOS

>      ```bash

Reviewer: We can write a small ballerina program to do this IMO, then we don't need multiple sessions for linux and windows.

Author: added a ballerina program to readme.md and package.md

     --url https://api.hubapi.com/oauth/v1/token \
     --header 'content-type: application/x-www-form-urlencoded' \
     --data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
     ```

   - Windows

     ```bash
     curl --request POST ^
     --url https://api.hubapi.com/oauth/v1/token ^
     --header 'content-type: application/x-www-form-urlencoded' ^
     --data 'grant_type=authorization_code&code=<CODE>&redirect_uri=<YOUR_REDIRECT_URI>&client_id=<YOUR_CLIENT_ID>&client_secret=<YOUR_CLIENT_SECRET>'
     ```

   This command will return the access token necessary for API calls.

   ```json
   {
     "token_type": "bearer",
     "refresh_token": "<Refresh Token>",
     "access_token": "<Access Token>",
     "expires_in": 1800
   }
   ```

5. Store the access token securely for use in your application.

## Quickstart

To use the `HubSpot Properties connector` in your Ballerina application, update the `.bal` file as follows:

### Step 1: Import the module

Import the `hubspot.properties` module and `oauth2` module.

```ballerina
import ballerinax/hubspot.properties as hsproperties;

> import ballerina/oauth2;

Reviewer: Why do we need to import `oauth2` module ?

Reviewer: In that case this is correct. But, since this is an example are there any other auth-schemas supported by the HubSpot which we can use here (without any additional imports) ?

Author: No, should I remove it ?

Reviewer 2: I am okay with having the auth import, otherwise we have to use a string value here, which isn't ideal.

```
### Step 2: Instantiate a new connector

1. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

> 2. Instantiate a `hsproperties:ConnectionConfig` with the obtained credentials and initialize the connector with it.

Reviewer: Shall we switch the order ?

Author: done

    ```ballerina 
    configurable string clientId = ?;
    configurable string clientSecret = ?;
    configurable string refreshToken = ?;

    final hsproperties:ConnectionConfig config = {
        auth : {
            clientId,
            clientSecret,
            refreshToken,
            credentialBearer: oauth2:POST_BODY_BEARER
        }
    };

>     final hsproperties:Client hsproperties = check new (config, "https://api.hubapi.com/crm/v3/properties");

Reviewer: AFAIU the `serverUrl` parameter has a default value. If so, do we need to specify the `serverUrl` like this ?

Author: no. Removed serviceUrl

    ```
### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Create a Property Group

```ballerina

> hsproperties:PropertyGroupCreate propertyGroupInput = { "name": "examplePropertyGroup",

Reviewer: Lets move this inside the `main` function.

Author: done

    configurable string refreshToken = ?;

    final hsproperties:ConnectionConfig config = {
        auth : {
            clientId,
            clientSecret,
            refreshToken,
            credentialBearer: oauth2:POST_BODY_BEARER
        }
    };

    final hsproperties:Client hsproperties = check new (config);
    ```

2. Create a `Config.toml` file and, configure the obtained credentials in the above steps as follows:

   ```toml
    clientId = <Client Id>
    clientSecret = <Client Secret>
    refreshToken = <Refresh Token>
   ```

### Step 3: Invoke the connector operation

Now, utilize the available connector operations. A sample usecase is shown below.

#### Create a Property Group

```ballerina
public function main() returns error? {
   hsproperties:PropertyGroupCreate propertyGroupInput = { "name": "examplePropertyGroup",
                                           "displayOrder": -1,
                                           "label": "This is an example Property Group"};
                                           
   hsproperties:PropertyGroup response = check hubspot->/[testObjectType]/groups.post(payload = propertyGroupInput);
}

> ```

Reviewer: Lets add a step to run the ballerina program into the `Quickstart` guide.

https://github.com/ballerina-platform/module-ballerinax-twitter/blob/main/ballerina/Module.md#step-4-run-the-ballerina-application

Author: done
